File: anime_audio_backend/db_models/generated_character.py
from sqlalchemy import Boolean, Column, DateTime, Integer, String, Text, ForeignKey
from sqlalchemy.sql import func

from database import Base


ART_STYLE_HINTS = {
    "modern_anime": "modern polished anime illustration, crisp digital rendering, detailed lighting, clean linework",
    "90s_anime": "1990s anime cel shading, retro OVA aesthetic, thicker outlines, softer film-like colors, hand-painted anime frame feel, less glossy rendering, classic 90s eye design",
    "80s_anime": "1980s anime aesthetic, bold cel shading, vintage palette, sharper facial structure, classic hand-drawn anime look",
    "early_2000s_anime": "early 2000s anime look, lighter digital coloring, classic TV anime rendering, simpler highlights",
    "retro_ova": "retro OVA anime frame, cinematic cel shading, painted background feel, nostalgic composition",
    "shojo_90s": "1990s shojo anime style, elegant large eyes, delicate linework, pastel color palette, romantic vintage anime feel",
    "grainy_vhs_anime": "retro anime VHS look, visible film grain, analog softness, faded colors, nostalgic tape-era anime frame",
    "cel_shaded_classic": "classic anime cel shading, flat shadow layers, hand-drawn look, bold outlines, traditional animation feel",
}


class GeneratedCharacter(Base):
    __tablename__ = "generated_characters"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(
        Integer,
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    prompt = Column(Text, nullable=False)
    style = Column(String(100), nullable=False)
    art_style = Column(String(100), nullable=False, default="modern_anime")

    gender = Column(String(50), nullable=False)
    hair = Column(String(50), nullable=False)
    eyes = Column(String(50), nullable=False)
    outfit = Column(String(50), nullable=False)
    mood = Column(String(50), nullable=False)

    image_url = Column(Text, nullable=False)
    is_favorite = Column(Boolean, default=False, nullable=False)

    name = Column(String(120), nullable=True)
    backstory = Column(Text, nullable=True)
    story_scene = Column(Text, nullable=True)

    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

File: lib/models/generated_character.dart
class GeneratedCharacter {
  static const String backendBaseUrl = 'https://anime-seek.com/fastapi';

  final int? id;
  final String prompt;
  final String style;
  final String artStyle;
  final String gender;
  final String hair;
  final String eyes;
  final String outfit;
  final String mood;
  final String imageUrl;
  final bool isFavorite;
  final DateTime? createdAt;

  final String? name;
  final String? backstory;
  final String? storyScene;

  GeneratedCharacter({
    this.id,
    required this.prompt,
    required this.style,
    this.artStyle = 'modern_anime',
    required this.gender,
    required this.hair,
    required this.eyes,
    required this.outfit,
    required this.mood,
    required this.imageUrl,
    this.isFavorite = false,
    this.createdAt,
    this.name,
    this.backstory,
    this.storyScene,
  });

  String get resolvedImageUrl {
    if (imageUrl.startsWith('http://') || imageUrl.startsWith('https://')) {
      return imageUrl;
    }

    if (imageUrl.startsWith('/fastapi/')) {
      return 'https://anime-seek.com$imageUrl';
    }

    if (imageUrl.startsWith('/')) {
      return '$backendBaseUrl$imageUrl';
    }

    return '$backendBaseUrl/$imageUrl';
  }

  GeneratedCharacter copyWith({
    int? id,
    String? prompt,
    String? style,
    String? artStyle,
    String? gender,
    String? hair,
    String? eyes,
    String? outfit,
    String? mood,
    String? imageUrl,
    bool? isFavorite,
    DateTime? createdAt,
    String? name,
    String? backstory,
    String? storyScene,
  }) {
    return GeneratedCharacter(
      id: id ?? this.id,
      prompt: prompt ?? this.prompt,
      style: style ?? this.style,
      artStyle: artStyle ?? this.artStyle,
      gender: gender ?? this.gender,
      hair: hair ?? this.hair,
      eyes: eyes ?? this.eyes,
      outfit: outfit ?? this.outfit,
      mood: mood ?? this.mood,
      imageUrl: imageUrl ?? this.imageUrl,
      isFavorite: isFavorite ?? this.isFavorite,
      createdAt: createdAt ?? this.createdAt,
      name: name ?? this.name,
      backstory: backstory ?? this.backstory,
      storyScene: storyScene ?? this.storyScene,
    );
  }

  factory GeneratedCharacter.fromJson(Map<String, dynamic> json) {
    return GeneratedCharacter(
      id: json['id'],
      prompt: json['prompt'] ?? '',
      style: json['style'] ?? '',
      artStyle: json['art_style'] ?? json['artStyle'] ?? 'modern_anime',
      gender: json['gender'] ?? '',
      hair: json['hair'] ?? '',
      eyes: json['eyes'] ?? '',
      outfit: json['outfit'] ?? '',
      mood: json['mood'] ?? '',
      imageUrl: json['image_url'] ?? '',
      isFavorite: json['is_favorite'] ?? false,
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'])
          : null,
      name: json['name'],
      backstory: json['backstory'],
      storyScene: json['story_scene'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'prompt': prompt,
      'style': style,
      'art_style': artStyle,
      'gender': gender,
      'hair': hair,
      'eyes': eyes,
      'outfit': outfit,
      'mood': mood,
      'image_url': imageUrl,
      'is_favorite': isFavorite,
      'created_at': createdAt?.toIso8601String(),
      'name': name,
      'backstory': backstory,
      'story_scene': storyScene,
    };
  }
}